src/k2/httpproxy/HTTPProxy.cpp

```diff
@@ -316,7 +234,38 @@ HTTPProxy::_handleRead(shd::ReadRequest&& request) {
 seastar::future<std::tuple<sh::Status, shd::QueryResponse>>
 HTTPProxy::_handleQuery(shd::QueryRequest&& request) {
     K2LOG_D(log::httpproxy, "Received query request {}", request);
-    return MakeHTTPResponse<shd::QueryResponse>(sh::Statuses::S501_Not_Implemented("query not implemented"), shd::QueryResponse{});
+    auto iter = _txns.find(request.timestamp);
+    if (iter == _txns.end()) {
+        return MakeHTTPResponse<shd::QueryResponse>(Txn_S410_Gone, shd::QueryResponse{});
+    }
+    auto queryIter = iter->second.queries.find(request.queryId);
+    if (queryIter ==iter->second.queries.end()) {
+        return MakeHTTPResponse<shd::QueryResponse>(sh::Statuses::S410_Gone("Query does not exist"), shd::QueryResponse{});
+    }
+
+    return iter->second.handle.query(queryIter->second)
+    .then([this, request=std::move(request)](QueryResult&& result) {
+        if(!result.status.is2xxOK()) {
+            return MakeHTTPResponse<shd::QueryResponse>(sh::Status{.code = result.status.code, .message = result.status.message}, shd::QueryResponse{});
+        }
+        std::vector<shd::SKVRecord::Storage> records;
+        records.reserve(result.records.size());
+        for (auto& k2record: result.records) {
+            sh::String collectionName(k2record.collectionName);
+            // k2 schema is already populated by query api, get corresponding shd schema from cache
+            auto shSChema = getSchemaFromCache(collectionName, k2record.schema);
+            auto rec = _buildSHDRecord(k2record, collectionName, shSChema);
+            records.push_back(std::move(rec.getStorage()));
+        }
+        auto& txnrec = _txns.at(request.timestamp);
+        auto& query = txnrec.queries.at(request.queryId);
```

Comment:
these lines can throw an exception. You should probably catch that and convert to some non-retryable error

Reply:
yes it can. Same reason as other instances

Reply:
Done